engine/src/flutter/shell/platform/darwin/ios/framework/Source/UITextInput+TextRange.swift

```diff
@@ -0,0 +1,226 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import UIKit
+
+@objc(FlutterTextPosition)
+final class TextPosition: UITextPosition, NSCopying {
+  let offset: UInt
+  @objc let affinity: UITextStorageDirection
+
+  init(index: UInt, affinity: UITextStorageDirection = .forward) {
+    self.offset = index
+    self.affinity = affinity
+  }
+
+  override func isEqual(_ object: Any?) -> Bool {
+    guard let other = object as? TextPosition else {
+      return false
+    }
+    // Affinity is only visual.
+    return offset == other.offset
+  }
+
+  func copy(with zone: NSZone? = nil) -> Any {
+    return TextPosition(index: offset, affinity: affinity)
+  }
+
+  override var description: String {
+    "TextPosition(\(index))"
+  }
+}
+
+@objc(FlutterTextRange)
+final class TextRange: UITextRange, NSCopying {
+  let nsRange: NSRange
+
+  init(NSRange range: NSRange) {
+    assert(range.location != NSNotFound)
+    self.nsRange = range
+  }
+
+  func copy(with zone: NSZone? = nil) -> Any {
+    TextRange(NSRange: nsRange)
+  }
+
+  override var start: TextPosition {
+    TextPosition(index: UInt(nsRange.lowerBound))
+  }
+
+  override var end: TextPosition {
+    TextPosition(index: UInt(nsRange.upperBound))
+  }
+
+  override var isEmpty: Bool { range.length == 0 }
+
+  override func isEqual(_ object: Any?) -> Bool {
+    guard let other = object as? TextRange else {
+      return false
+    }
+    return range == other.nsRange
+  }
+
+  /// Returns the smallest TextRange that contains this TextRange and doesn't contain unpaired surrogates.
```

Comment:
can you add comment on how it's different from `rangeOfComposedCharacter` with the emoji example for future readers. it's very hard to understand without research in unicode.

Reply:
Again I think the documentation here per se seems sufficient (I thought "unpaired surrogate" is a bit of a made-up term but it's actually used in the wikipedia page of UTF-16 too). A README / contributing guide is probably a better place to cover basic unicode / string encoding concepts?

Reply:
(I'll create a README in a different PR).